internal/stolonctl/pkg/cluster/upgrade.go

```
@@ -53,12 +63,65 @@ func Upgrade(ctx context.Context, config Config) error {
 		Name:      defaults.CRDName,
 		Namespace: config.Namespace,
 	}
-	res, err := crdclient.Create(ctx, objMeta)
-	if err != nil && !trace.IsAlreadyExists(err) {
+	res, err := crdclient.CreateOrRead(objMeta)
+	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	// temporary
 	log.Info(res)
+	res, err = crdclient.MarkStepCompleted(res, crd.StolonUpgradeStepInit)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Backup Postgres
+	res, err = crdclient.MarkStepStarted(res, crd.StolonUpgradeStepBackupPostgres)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err = backupPostgres(config); err != nil {
+		return trace.Wrap(err)
+	}
+	res, err = crdclient.MarkStepCompleted(res, crd.StolonUpgradeStepBackupPostgres)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func backupPostgres(config Config) error {
+
+	if err := createPgPassFile(config); err != nil {
+		return trace.Wrap(err)
+	}
+	options := []string{fmt.Sprintf("-h%s", config.PostgresHost),
+		fmt.Sprintf("-p%s", config.PostgresPort),
+		fmt.Sprintf("-U%s", config.PostgresUser), "-c",
+		fmt.Sprintf("-f%s", config.PostgresBackupPath)}
+
+	_, err := exec.Command(pgDumpCommand, options...).CombinedOutput()
+	if err != nil && !isExitError(err) {
```

You have to check the exit code - otherwise you're ignoring errors.